core/shared/src/main/scala/fs2/internal/Algebra.scala

```diff
@@ -147,8 +138,24 @@ private[fs2] object Algebra {
       if (midAcquires == 0) midAcquiresDone()
     }
 
-    def releaseResource(t: Token): Option[F[Unit]] = monitor.synchronized {
-      resources.remove(t).map(_.forceRelease)
+    /**
```

Comment:
the implementation is not stack safe. 
Also michael I think we have to clarify how the scopes and spawns work really.

Reply:
Any recommendations for fixing stack safety? We could trampoline through F I suppose.